core/src/main/java/io/lionweb/model/impl/AbstractClassifierInstance.java

```diff
@@ -44,35 +44,47 @@ public abstract class AbstractClassifierInstance<T extends Classifier<T>>
   }
 
   @Override
-  public void addAnnotation(@Nonnull AnnotationInstance instance) {
+  public boolean addAnnotation(@Nonnull AnnotationInstance instance) {
     Objects.requireNonNull(instance);
     if (this.annotations == null) {
       this.annotations = new ArrayList<>();
     }
-    if (this.annotations.contains(instance)) {
+    if (instance.getID() != null
+        && annotations.stream().anyMatch(a -> a.getID().equals(instance.getID()))) {
       // necessary to avoid infinite loops and duplicate insertions
-      return;
+      return false;
     }
     if (instance instanceof DynamicAnnotationInstance) {
       ((DynamicAnnotationInstance) instance).setAnnotated(this);
     }
-    if (this.annotations.contains(instance)) {
-      // necessary to avoid infinite loops and duplicate insertions
-      // the previous setAnnotated could potentially have already set annotations
-      return;
+    if (instance.getID() == null
+        || annotations.stream().noneMatch(a -> a.getID().equals(instance.getID()))) {
+      this.annotations.add(instance);
+      if (observer != null) {
+        observer.annotationAdded(this, this.annotations.size() - 1, instance);
+      }
     }
-    this.annotations.add(instance);
+    return true;
   }
 
   @Override
-  public void removeAnnotation(@Nonnull AnnotationInstance instance) {
+  public int removeAnnotation(@Nonnull AnnotationInstance instance) {
     Objects.requireNonNull(instance);
-    if (annotations == null || !this.annotations.remove(instance)) {
+    int index = -1;
+    if (annotations != null) {
+      index = this.annotations.indexOf(instance);
+    }
+    if (index == -1) {
       throw new IllegalArgumentException();
     }
+    annotations.remove(index);
     if (instance instanceof DynamicAnnotationInstance) {
       ((DynamicAnnotationInstance) instance).setAnnotated(null);
     }
+    if (observer != null) {
+      observer.annotationRemoved(this, index, instance);
+    }
+    return index;
   }
 
   void tryToRemoveAnnotation(@Nonnull AnnotationInstance instance) {
```

Comment:
Why does this not return an `int`, or `bool`, or maybe `Integer` (with `null` meaning "not removed")?

Reply:
(the comment is on the interface from which it is overridden)